spark3/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java

```diff
@@ -101,24 +103,35 @@ public Table getTable(StructType schema, Transform[] partitioning, Map<String, S
     SparkSession spark = SparkSession.active();
     setupDefaultSparkCatalog(spark);
     String path = options.get("path");
+    Long snapshotId = Spark3Util.propertyAsLong(options, SparkReadOptions.SNAPSHOT_ID, null);
+    Long asOfTimestamp = Spark3Util.propertyAsLong(options, SparkReadOptions.AS_OF_TIMESTAMP, null);
```

Comment:
I think that modifications to the source for snapshot and timestamp are unnecessary. Looking at the Spark code, the table will be loaded using the [catalog/identifier from `SupportsCatalogOptions`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala#L114-L120) and then [wrapped in a `DataSourceV2Relation` that uses the reader options](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala#L131). Those read options still include `snapshot-id` and `as-of-timestamp`.

When a scan is created, those [options get passed into the scan builder](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala#L42-L45).

There should be no need to handle these in `IcebergSource` because the options will be set at planning time. Not considering them here avoids the need to add `SnapshotAwareIdentifier` and removes ambiguous cases where the scan options identify a snapshot and the identifier did as well.

Can you update to remove this?

Reply:
Let me look into this.

Reply:
It turns out that you are mistaken.
The [`DataSourceV2Relation`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala#L131) is [created](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala#L175-L176) with the output attributes from the schema given by `SparkTable#schema()`.
The `SparkTable` is loaded using the catalog and identifier, and that is why I need the `SnapshotAwareIdentifier` when loading it, so that I can return the snapshot schema in `SparkTable#schema()`.
Otherwise the `SparkScanBuilder` does have the options for `snapshot-id` or `as-of-timestamp`, but its `pruneColumns()` will be called by Spark with a `requestedSchema` that is a subset of the table schema. Then its `build()` will return a `SparkBatchQueryScan` with an incorrect schema.
Once I remove the modifications to `IcebergSource`, `SparkCatalog` and `SparkTable`, then the unit tests I added all fail, as I suspected.

Reply:
Great, thanks for looking into it more! I've been considering other options to solve this problem and I think we probably will need to go with something similar to what you have here.

The output attributes problem that you found makes sense. If there are columns in the snapshot's schema that have been dropped, then Spark doesn't have attributes for those columns and can't project them. We could try to expose the deleted columns as metadata columns in 3.1 and later, but we would still have a problem with `*` expansion because that uses the initial set of columns reported by the table and not metadata columns. So `select *` wouldn't behave the same way when time traveling if we took the approach with metadata columns.

We could also try to solve this on the Spark side, but I don't think that it's unreasonable for Spark to do what it is here and expect that the table reports a schema initially that can be used for query analysis.

The option that's left is loading the correct snapshot of the table so I agree with this direction. I'm weighing whether this should be done like metadata tables in the Iceberg layer or above it in Spark like you have here.